### ISSUE TITLE
[BUG] Reporting ENOTEMPTY (-39) on any installation. (Folder is nonexistant)

### ISSUE THREAD

**Issue by Nebrethar** [npm/cli#1119]
# What / Why
I have been toying around with uPort's libraries, but it looks like my NPM is broken as a result.

## When
  * Happens when I try to install anything with NPM (See code snippet toward the bottom)

## Where
  * NPM, obtained from running `sudo apt-get install npm`
  * Ubuntu Desktop 18.04

### Current Behavior
Anytime I try to install anything, it will throw an error related to a package folder that does not exist. 
__My npm has also gained the ability to delete a package, then throw this error during the reinstall.__ 

### Steps to Reproduce
Here is my command history for the time i've been experiencing the error:
__NOTE:__ Before this, I was only editing JS files. [Here is what I was looking at and fixing](https://stackoverflow.com/questions/46176973/es6-import-statements-not-working-in-node-v8-4-0)
```
  515  nano boilerplate.js
  516  node boilerplate.js
  517  nano connect.js
  518  nano boilerplate.js
  519  nano connect.js
  520  node boilerplate.js
  521  nano boilerplate.js
  522  node boilerplate.js
  523  nano boilerplate.js
  524  node boilerplate.js
  525  npm install uport-connect
  526  sudo npm install uport-connect
  527  sudo mv /home/neb/Projects/uport/node_modules/express /home/neb/Projects/uport/node_modules/express.bak
  528  node boilerplate.js
  529  sudo npm install uport-connect
  530  nano /home/neb/Projects/uport/node_modules/express
  531  cd /home/neb/Projects/uport/node_modules
  532  ls
  533  npm install express
  534  sudo npm install express
  535  ls
  536  sudo npm install uport-connect
  537  sudo mv -r /home/neb/Projects/uport/node_modules/ethr-did-resolver /home/neb/Projects/uport/node_modules/ethr-did-resolver.bak
  538  sudo mv /home/neb/Projects/uport/node_modules/ethr-did-resolver /home/neb/Projects/uport/node_modules/ethr-did-resolver.bak
  539  sudo npm install uport-connect
  540  npm audit
  541  npm audit fix
  542  npm install npm-audit
  543  npm restart
  544  sudo apt-get remove --purge npm
  545  sudo apt-get install npm
  546  sudo npm install uport-connect
  547  cat /home/neb/Projects/uport/node_modules/ethr-did-resolver
```

### Expected Behavior

I expected npm to remove and install packages fluidly, as it normally does 😄 

## Who
* n/a

## References
Error Text:
```
npm ERR! Linux 5.3.0-45-generic
npm ERR! argv "/usr/bin/node" "/usr/bin/npm" "install" "uport-connect"
npm ERR! node v8.10.0
npm ERR! npm  v3.5.2
npm ERR! path /home/***/Projects/uport/node_modules/ethr-did-resolver
npm ERR! code ENOTEMPTY
npm ERR! errno -39
npm ERR! syscall rename

npm ERR! ENOTEMPTY: directory not empty, rename '/home/***/Projects/uport/node_modules/ethr-did-resolver' -> '/home/***/Projects/uport/node_modules/.ethr-did-resolver.DELETE'
npm ERR!
npm ERR! If you need help, you may report this error at:
npm ERR!     <https://github.com/npm/npm/issues>

npm ERR! Please include the following file with any support request:
npm ERR!     /home/***/Projects/uport/node_modules/npm-debug.log
(uport3) ***@***:~/Projects/uport/node_modules$ cat /home/***/Projects/uport/node_modules/ethr-did-resolver
```
Result of cat on the location:
```
(uport3) ***@***:~/Projects/uport/node_modules$ cat /home/***/Projects/uport/node_modules/ethr-did-resolver
cat: /home/***/Projects/uport/node_modules/ethr-did-resolver: No such file or directory
```

**Comment by ljharb** [comment]
Don’t install npm by itself; npm comes with node.

**Comment by SMotaal** [comment]
@Nebrethar, do you want to first confirm if installing `nodejs` does work? 

I think it is fair to want to make sure if what I presumed below is the case first and foremost.

---

@ljharb Sure, `thing install nodejs` is the more usual way for wanting to `npm install things`, but it makes sense to presume that `thing install npm` should work…

So even it does that by asking… `we did not find node, which is needed, we can do that, or you can do it and come back, should we do it for you?`… that not being the case is not because `thing install npm` is not right!

**Comment by Nebrethar** [comment]
Here is a fresh install of node. It did not come with npm. But it did tell me to use `sudo apt install npm`
```
neb@neb:~$ sudo apt-get install nodejs
Reading package lists... Done
Building dependency tree
Reading state information... Done
The following packages were automatically installed and are no longer required:
  gyp javascript-common libjs-async libjs-inherits libjs-jquery libjs-node-uuid libjs-underscore libssl1.0-dev libuv1-dev
Use 'sudo apt autoremove' to remove them.
The following NEW packages will be installed:
  nodejs
0 upgraded, 1 newly installed, 0 to remove and 5 not upgraded.
Need to get 0 B/4,796 kB of archives.
After this operation, 18.5 MB of additional disk space will be used.
Selecting previously unselected package nodejs.
(Reading database ... 163240 files and directories currently installed.)
Preparing to unpack .../nodejs_8.10.0~dfsg-2ubuntu0.4_amd64.deb ...
Unpacking nodejs (8.10.0~dfsg-2ubuntu0.4) ...
Setting up nodejs (8.10.0~dfsg-2ubuntu0.4) ...
update-alternatives: using /usr/bin/nodejs to provide /usr/bin/js (js) in auto mode
Processing triggers for man-db (2.8.3-2ubuntu0.1) ...
neb@neb:~$ npm

Command 'npm' not found, but can be installed with:

sudo apt install npm
```

**Comment by Nebrethar** [comment]
I reinstalled node and reinstalled npm. The problem seemed fixed for a bit, but I ran into the same error with a different package:

```npm ERR! Linux 5.3.0-45-generic
npm ERR! argv "/usr/bin/node" "/usr/bin/npm" "install" "ngrok"
npm ERR! node v8.10.0
npm ERR! npm  v3.5.2
npm ERR! path /home/neb/Projects/uport/node_modules/ethr-did-resolver
npm ERR! code ENOTEMPTY
npm ERR! errno -39
npm ERR! syscall rename

npm ERR! ENOTEMPTY: directory not empty, rename '/home/neb/Projects/uport/node_modules/ethr-did-resolver' -> '/home/neb/Projects/uport/node_modules/.ethr-did-resolver.DELETE'
npm ERR!
npm ERR! If you need help, you may report this error at:
npm ERR!     <https://github.com/npm/npm/issues>

npm ERR! Please include the following file with any support request:
npm ERR!     /home/neb/Projects/uport/expressJS/npm-debug.log
```

I'm starting to think this problem is related to anytime I install uport-connect. This may be an issue for them...

![image](https://user-images.githubusercontent.com/34951354/78675803-31e42f00-78ab-11ea-99e2-f8125c55604e.png)

UPDATE: I used an Ubuntu 18.04 VM to test the theory that the issue may be more relevant to UPort's software. I was able to install all of these packages without an issue. I will try to recreate the issue today. 

See Screenshot...

**Comment by ljharb** [comment]
@Nebrethar i mean, you shouldn't install node or npm with apt regardless; whoever maintains that apt repository was incorrect for distributing npm independently.

The error at the end of your screenshot is that you're missing a package.json file.

**Comment by Nebrethar** [comment]
@ljharb The screenshot at the end of my post was to signify an installation that worked for me. That's all it was there for 😅

I found out that I can fix this issue by running this command, with no `sudo`, every time before I install a package:
`rm -Rf /home/neb/Projects/uport/node_modules`

The code issue is that anytime I installed something, npm would create a hidden file with the extra extension `.DELETE` with every change to a node module. 

Right now, I have to remove every file ending in `.DELETE` in my node modules folder, every time I run an installation for anything. 

I don't think that's a result of installing it via apt, as I have used the `apt-get` version on Ubuntu for years and never had that issue.
